File: src/main/java/com/IceCreamQAQ/Yu/controller/DefaultActionInvoker.java
package com.IceCreamQAQ.Yu.controller;

import com.IceCreamQAQ.Yu.entity.DoNone;
import com.IceCreamQAQ.Yu.entity.Result;
import lombok.val;

@Deprecated
public class DefaultActionInvoker extends ActionInvoker {
    @Override
    public void invoke(String path, ActionContextBase context) throws Exception {
        try {
            for (MethodInvoker before : befores) {
                before.invoke(context);
            }

            val reObj = invoker.invoke(context);

            context.setSuccess(true);

            if (reObj == null) context.setResult(null);

            if (reObj instanceof String) context.buildResult((String) reObj);
            else if (reObj instanceof Result) context.setResult((Result) reObj);
            else context.buildResult(reObj);

        } catch (DoNone ignored) {
        } catch (Result result) {
            context.setResult(result);
        }
    }
}

File: src/main/kotlin/com/IceCreamQAQ/Yu/controller/router/DefaultActionInvoker.kt
package com.IceCreamQAQ.Yu.controller.router

import com.IceCreamQAQ.Yu.controller.ActionContext
import com.IceCreamQAQ.Yu.entity.DoNone
import com.IceCreamQAQ.Yu.entity.Result
import java.lang.Exception

@Deprecated("已经弃用")
open class DefaultActionInvoker(level: Int) : DefaultRouter(level) {

    lateinit var befores: Array<MethodInvoker>
    lateinit var invoker: MethodInvoker

    override fun invoke(path: String, context: ActionContext): Boolean {
        if (super.invoke(path, context)) return true
        try {
            for (before in befores) {
                val o = before.invoke(context)
                if (o != null) context[toLowerCaseFirstOne(o::class.java.simpleName)] = o
            }
            val result = invoker.invoke(context) ?: return true
            if (result is Result) context.result = result
            else context.result = context.buildResult(result)
        } catch (e: DoNone) {
        } catch (e: Result) {
            context.result = e
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return true
    }

    private fun toLowerCaseFirstOne(s: String): String {
        return if (Character.isLowerCase(s[0])) s
        else (StringBuilder()).append(Character.toLowerCase(s[0])).append(s.substring(1)).toString();
    }
}